Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp

```diff
@@ -119,9 +123,20 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
     callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
         resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
     });
+
+    // To reduce fingerprinting, if the origin has not requested permission to use the
+    // Notifications API, and the permission state is "denied", return "default" instead.
+    if (resultPermission == NotificationClient::Permission::Denied && !m_notificationPermissionRequesters.contains(context->securityOrigin()->data()))
+        return NotificationClient::Permission::Default;
+
```

Comment:
I would leave `WebNotificationClient::checkPermission` alone because within WebCore and WebKit, we need to know the true permission state and not the downgraded version of the permission state. For instance, `PushManager::subscribe` will check the permission state, and if the permission state is `Default`, it will show a permission dialog. With this change as-is, `Denied` will get mapped to `Default` in every new WebProcess. This means that users will see a permission prompt for origins where they've actively denied permissions within the browser settings pane. This is incorrect behavior.

You can add some other function that returns the downgraded permission state and have only `Notification::permission` and `PushManager::permissionState` call that new function.

Reply:
I think we should test it but I personally don't think we have a bug. I think it is working as intended. I don't think the statement about double prompting is correct.

If the user previously denied, then WebCore thinks the permissions is "prompt" as is intended for privacy and will think it needs to requestPermission. It will request permission but it will NOT prompt since the UIProcess/client knows the policy is denied, it will simply reject the permission request.

This is the intended behavior of this code change for notification and the same should apply to push. Otherwise, we will end up with ways for fingerprint the notification permission state via the push API instead.

Again, we should check it but I am not convinced a code change will be needed.

Reply:
> If the user previously denied, then WebCore thinks the permissions is "prompt" as is intended for privacy and will think it needs to requestPermission. It will request permission but it will NOT prompt since the UIProcess/client knows the policy is denied, it will simply reject the permission request.

The current implementation of the request permission callback our browser doesn't follow this rule (see the implementation of the `_webView:requestNotificationPermissionForSecurityOrigin:decisionHandler:` delegate callback). It will unconditionally show a prompt if the current decision is not `UserNotificationPermissionPolicy::Allow`.

So I think if you want to commit this patch as-is, you need a follow-up patch in the browser to fix this behavior, or else users will get gratuitous permission prompts for Web Push when they've already actively denied that permission.

Reply:
Even though we will call requestNotificationPermissionForSecurityOrigin again, Safari will NOT prompt again if it knows the answer (allow or deny). I still think we should verify manually before landing but I am fairly confident the patch will work as-is.

Reply:
Oops nevermind, I misread that part of the browser code. Looks like we call the decision handler in the browser properly when the permission is set to `UserNotificationPermissionPolicy::Deny`. So you're good to commit this.

Reply:
Sounds good--I didn't make any changes besides fixing the API-test failure.